### PR TITLE
adjusting links to enable pixel, wavelength linking in specviz2d

### PR DESCRIPTION
**Pull request by gibsongreen**
This pull request is to address subset linking in specviz2d [:cat:](https://jira.stsci.edu/browse/JDAT-3508).

**Description:**
This update updates LinkSame to LinkSameWithUnits, and appropriate components depending on the spectrum dimensionality is handled so subsets load within the bounds of both viewers, irregardless of which they were created in, nor if additional data items are added to the data collection (eg. via Spectral Extraction Plugin). 

These updates should also ensure that unit handling for subsets is occurring. Unit changes can be seen in the plugin.

Note:
- If using `specviz2d.app.get_subsets(use_display_units=True)` needs the use_display_units argument. 
 

**Bug Behavior**:

https://github.com/spacetelescope/jdaviz/assets/42986583/84cd834a-6827-460b-adf0-0751f29930af

**Updated Behavior**:

https://github.com/user-attachments/assets/ecc0c45b-b83a-4640-96fb-5fe7fc343564

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by javerbukh** [comment]
I was able to draw subset 1 on the spectrum viewer and subset 2 on the 2d viewer and they appear at the right locations, but the units for subset 2 show up as `um` when I run `specviz2d.app.get_subsets()` instead of `pix`.

image

Looks like the reason for that is the `get_subsets` method assumes an xrange roi is spectral and thus applies the spectral units to it. We may need to add a check if the spectral subset was created on a 2d viewer or spectrum viewer.

https://github.com/spacetelescope/jdaviz/blob/c731eef855a0a65a6172fe53fca499ff6a98b3d9/jdaviz/app.py#L943-L945

**Comment by kecnry** [comment]
> but the units for subset 2 show up as um when I run specviz2d.app.get_subsets() instead of pix.

I thought we had decided that was the intended behavior, at least for now until we have API support for requesting specific units?

**Comment by kecnry** [comment]
ah, sorry, I missed the screenshot.  Yes, it should show the actual wavelength values (along with wavelength units), but definitely not mixed with wavelength values and pixel units.

**Comment by gibsongreen** [comment]
> I was able to draw subset 1 on the spectrum viewer and subset 2 on the 2d viewer and they appear at the right locations, but the units for subset 2 show up as `um` when I run `specviz2d.app.get_subsets()` instead of `pix`.
> 
> image
> 
> Looks like the reason for that is the `get_subsets` method assumes an xrange roi is spectral and thus applies the spectral units to it. We may need to add a check if the spectral subset was created on a 2d viewer or spectrum viewer.
> 
> https://github.com/spacetelescope/jdaviz/blob/c731eef855a0a65a6172fe53fca499ff6a98b3d9/jdaviz/app.py#L943-L945

I meant to get back sooner I was deep in testing-land. μm, the unit is appearing in Subset Tools irregardless of what viewer the subset is drawn in. So the correct value is showing up independent of the viewer, but the units are always appearing in μm. 

In the case of the screenshot, if Subset 1 was drawn in the 1d viewer, and Subset 2d in the 2d viewer, are the values incorrect?

**Comment by gibsongreen** [comment]
Also, I split up the test into two temporarily. I'm going to have a look over the weekend. Drawing in 1d is working as expected for the test. Drawing in 2d, I am able to use the functionality in the test in a notebook, and it get the expected results. 

When I load the image from the spectrum in the test, the image is loading with 0s, and the shape of the image is not as it should be.

**Comment by kecnry** [comment]
Getting the spectral display units is not the right assumption here if the subsets can be returned in pixels, but if that was the behavior in main, then can probably be considered a follow-up effort (please create a ticket).

File: jdaviz/tests/test_subsets.py [hunk 1]
```diff
@@ -791,6 +792,94 @@ def test_only_overlapping_in_specviz2d(specviz2d_helper, mos_spectrum2d):
...
+
+    specviz2d_helper.load_data(spectrum_2d=spectrum_data)
+    viewer_1d = specviz2d_helper.app.get_viewer(
+        specviz2d_helper._default_spectrum_viewer_reference_name)
+    viewer_2d = specviz2d_helper.app.get_viewer(
+        specviz2d_helper._default_spectrum_2d_viewer_reference_name)
+
+    # subset drawn in 1d viewer, want data in 2d viewer
+    viewer_1d.apply_roi(XRangeROI(.0001, .0002))
+    subset_drawn_1d = viewer_2d.native_marks[-1].image
```
Comment (gibsongreen) [inline jdaviz/tests/test_subsets.py:867]: Getting the 3d image array this way is loading incorrectly only in the test. It should be (155, 864, 4) [which is how it loads in a Notebook], in the test file it is loading a zero array of shape (10,10,4), so the test thinks there's no mask. This is the last hold up.

**Comment by gibsongreen** [comment]
> Getting the spectral display units is not the right assumption here if the subsets can be returned in pixels, but if that was the behavior in main, then can probably be considered a follow-up effort (please create a ticket).

Will do right now!

**Comment by javerbukh** [comment]
Do you have an idea of why the python 3.11 CI test is failing?

**Comment by pllim** [comment]
Looks like this patch broke unit handling somewhere.
```
pext.export_bg_sub(add_data=True)

specviz2d/plugins/spectral_extraction/tests/test_spectral_extraction.py:170

astropy.units.core.UnitConversionError: 'um' (length) and '' (dimensionless) are not convertible
```

**Comment by kecnry** [comment]
The dev tests are failing everywhere and we have a tech debt ticket to fix that ASAP [:cat:](https://jira.stsci.edu/browse/JDAT-4340)

**Comment by pllim** [comment]
Are you saying this is affecting main? I restarted an old successful job on main from an hour ago. Let's see.

https://github.com/spacetelescope/jdaviz/actions/runs/8286222588

**Comment by kecnry** [comment]
I thought it was affecting other PRs, but maybe not, I do see it passing on others.  And since its specviz2d spectral extraction, the bug could be here...

**Comment by pllim** [comment]
The job on main still passes, so I think failure is related to this PR and cannot be ignored.

**Comment by gibsongreen** [comment]
> Are you saying this is affecting main? I restarted an old successful job on main from an hour ago. Let's see.
> 
> https://github.com/spacetelescope/jdaviz/actions/runs/8286222588

I am going through it with changed my local env to 3.11

**Comment by gibsongreen** [comment]
> The job on main still passes, so I think failure is related to this PR and cannot be ignored.

I am going through it now my env set to 3.11

**Comment by gibsongreen** [comment]
> Not able to have subsets appear in the other viewer. On a fresh environment in the Specviz2D notebook. image

@javerbukh is the the 1D spectrum auto-extracted from the 2d spectrum with the data you used or did you load the 1D and 2D separately using` load_data()`?

File: jdaviz/tests/test_subsets.py [hunk 2]
```diff
@@ -740,6 +742,104 @@ def test_only_overlapping_in_specviz2d(specviz2d_helper, mos_spectrum2d):
...
+    tolerance1 = 1e-6
+    expected_min1 = 6.e-5
+    expected_max1 = 8.e-05
+
+    assert np.allclose(min_value_subset1, expected_min1, atol=tolerance1)
+    assert np.allclose(max_value_subset1, expected_max1, atol=tolerance1)
+
+
+'''
+def test_draw1d_linking_specviz2d(specviz2d_helper):
```
[the remark this answers is not in the document]
Comment (gibsongreen) [reply jdaviz/tests/test_subsets.py:791]: That test is intentional commented out for now until the [SubsetCreateMessage PR](https://github.com/spacetelescope/jdaviz/pull/3238#issuecomment-2613115549) is merged (which I believe we got an update this morning that it will be occurring soon). We should test that subsets created in either viewer loaded in the opposing viewer as well (but this at the moment is prevented with the Glue maxpin)

**Review by javerbukh — APPROVED** [review]
The bug I was encountering was because I unsuccessfully loaded MAST data into the viewer (the URI would error and no data would be retrieved) and then tried to load local data using the import button without restarting the kernel. @gibsongreen is investigating this workflow but otherwise this works well!

File: CHANGES.rst [hunk 3]
```diff
@@ -747,6 +749,8 @@ Specviz
   display label for the y axis in spectral viewer. Previously it was hard coded
   to always display ``flux density`` no matter the input unit. [#2703]
 
+Specviz2d
+^^^^^^^^^
```
Comment (kecnry) [inline CHANGES.rst:753]: was this just from a remilestone and can be removed?

File: jdaviz/app.py [hunk 4]
```diff
@@ -766,8 +766,22 @@ def _link_new_data(self, reference_data=None, data_to_be_linked=None):
...
+                    # extracted spectrum added to data collection
+                    ref_wavelength_component = ref_data.components[3]
+                    links += [LinkSameWithUnits(linked_data.components[0], ref_data.components[1])]
+
+                links += [LinkSameWithUnits(linked_data.components[0], ref_data.components[0]),
+                          LinkSameWithUnits(linked_data.components[1], ref_wavelength_component)]
+            else:
+                links = [LinkSame(linked_data.components[0], ref_data.components[0]),
+                         LinkSame(linked_data.components[1], ref_data.components[1])]
+
```
Comment (kecnry) [inline jdaviz/app.py:784]: is there a more general way to do this by checking the component names themselves?  If its not straightforward, we can probably defer for future generalizations in the deconfigging effort.... I just worry this keeps adding to the hardcoded complexity that makes it hard to follow/debug the linking logic.

File: jdaviz/tests/test_subsets.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/tests/test_subsets.py:791]: can this be rebased and uncommented now?

File: jdaviz/app.py (hunk 4 above)
Comment (gibsongreen) [reply jdaviz/app.py:784]: I had thought about looping over both the `linked_data.components` and `ref_data_components` and determine which components to link. They are just the string names for the components so I don't think it be that expensive, but I was trying to avoid 2 loops here.

I was also basing the syntax off of the other links which do mostly using indexing to find the components, and I do very much agree readability/debugging is more challenging with using indexing.

 I don't mind making the adjustment, but I think it would also be worthwhile to have a follow up effort where we do this for all the other links.

File: CHANGES.rst (hunk 3 above)
Comment (gibsongreen) [reply CHANGES.rst]: It was, thought I had removed it once but that is now has been removed

File: jdaviz/app.py (hunk 4 above)
Comment (kecnry) [reply jdaviz/app.py:784]: Agreed - if it isn't something super straightforward, let's defer

**Review by kecnry — APPROVED** [review]
Thanks!
